pymodbus/client/base.py

```diff
@@ -143,15 +143,17 @@ def execute(self, request: ModbusRequest | None = None):
         :returns: The result of the request execution
         :raises ConnectionException: Check exception text.
         """
-        if not self.transport:
+        if not self.transport and self.reconnect_task is None:
             raise ConnectionException(f"Not connected[{self!s}]")
         return self.async_execute(request)
 
     # ----------------------------------------------------------------------- #
     # Merged client methods
     # ----------------------------------------------------------------------- #
-    async def async_execute(self, request) -> ModbusResponse:
+    async def _async_execute(self, request) -> ModbusResponse:
         """Execute requests asynchronously."""
+        if self.reconnect_task is not None:
```

Comment:
If there are not connection the APP should be notified, so that it can decide to stop reading data for a while.

Maybe a stupid example, but my solar inverter closes down every night between a couple of minutes uptown a couple of hours. So reconnect_max is 4 hours. The APP reads a single value every 50ms, think of the backlog even after a couple of minutes. A backlog bigger than just a few entries will cause problems with a lot of devices, because they typically reset the connection when receiving a burst of requests.

Reply:
> I mean, the APP will get an exception, but this will give reasonable behavior as the queue processing with effectively halt any further reads until the reconnect_task completes.

Well you blocked the exception, so basically you will just be queuing endlessly.

> Is that like a full transport loss or does it just stop responding to reads?

It disconnects, because it reboots.

> I'm not sure how this APP works...but it would have to be doing something very wrong for the queue to grow unbounded, a normal read loop will await the read(and thus will be blocked until the reconnect task completes if disconnected), a read loop should not create unbounded tasks.

In case of the disconnect (without notifying the APP) a reconnect task is started, and the queue just keeps growing until the inverter accepts a connect, once the connect is accepted, the worker_task will overload the inverter with requests leading to a close of the connection....this continues.

And yes the APP issues several reads independently and then uses async.gather to get the results later.

The whole point of the queue is to allow the APP to make multiple requests in parallel so of course the queue will grow when reconnecting.

Reply:
> Well you blocked the exception, so basically you will just be queuing endlessly.

But we do a `await self.reconnect_task`, which will block the read loop and prevent it from queuing endlessly.

> It disconnects, because it reboots.

Is the transport serial or tcp?

> And yes the APP issues several reads independently and then uses async.gather to get the results later.

But you await the gather right which would block until reconnected? So you are still bounded to the amount of independent reads in the gather.

> The whole point of the queue is to allow the APP to make multiple requests in parallel so of course the queue will grow when reconnecting.

Well its purpose is more to safely allow multiple tasks to use the same client, sure it's possible to indefinitely queue requests but I don't see a normal usage pattern that should cause something like that to happen, the queue should generally be bounded however to say the amount of tasks or amount of requests in the awaited gather.

Reply:
> But we do a `await self.reconnect_task`, which will block the read loop and prevent it from queuing endlessly.
> 
As I see the code in the pull request, sure it enters an await....but that means the request is not terminated but pending, which again means an endless number of requests pending the await.

> > It disconnects, because it reboots.
> 
> Is the transport serial or tcp?

For the sake of this discussion it is serial....since this PR is about serial

actually it is tcp to a serial converter placed next to the inverter but tomorrow I might replace the converter with my RPI4 server.

> But you await the gather right which would block until reconnected? So you are still bounded to the amount of independent reads in the gather.

There is a task that creates a request every 50ms, and adds the future to a dynamic list of futures, this is the collection process. 

When the analysis starts it syncs the requests by calling async.gather with the list as it is at that point, but the collection process continues.

Anyhow it is a bigger problem if the APP is not informed about a no-connection state, this goes hand in hand with the APP reconnect_callback at transport layer. The combination allows the APP to pause and resume requests. Without this the queue needs to:
- gather for requests limitations (bursts) in the device.
- remove duplicate requests (issues due to lack of response)
- set an upper limit to ensure the queue is not limitless.

> Well its purpose is more to safely allow multiple tasks to use the same client, sure it's possible to indefinitely queue requests but I don't see a normal usage pattern that should cause something like that to happen, the queue should generally be bounded however to say the amount of tasks or amount of requests in the awaited gather.

We do not define how the APP works, so we need to protect against patterns we have identified.

I think this PR have a lot of merit, but the issue of not reporting no-connecting to the APP is a big concern for me (based on experience).

Reply:
> There is a task that creates a request every 50ms, and adds the future to a dynamic list of futures, this is the collection process.

This is giving off major anti-pattern vibes to me. Like if you are reading from a few different registers...you should never make a request to a particular register unless there are no pending identical reqeusts to that same register.

> When the analysis starts it syncs the requests by calling async.gather with the list as it is at that point, but the collection process continues.

So it endlessly creates tasks which than call asyncio.gather on futures? This doesn't sound normal to me, is this code open source somewhere btw? Its somewhat unclear what exactly is going on here since it doesn't sound right to me.

Like I think you could see a similar sort of issue happening with entirely different protocols with this design pattern, for example if you made unbounded requests to a normal http endpoint that say couldn't processes requests fast enough to prevent all your system queues/buffers from filling up...you would see overloading issues, memory leaks and responsiveness issue's due to deep queue depths that would somewhat mirror their modbus counterparts.

I think this is more of a don't footgun yourself by designing your app that will leak resources like this, keep your buffers and queues small/bounded otherwise you might get a [bufferbloat](https://en.wikipedia.org/wiki/Bufferbloat) like affect. I don't see an obvious way of fixing this other than fixing how the APP uses the API.

The purpose of this queue mechanism isn't really to encourage someone to have a large amount/unbounded number of in flight requests so much as make it as safe as possible if it happens and to ensure that the client can be shared among tasks properly without requests/responses getting out of sync. A common design pattern I would expect to be used here would have a tasks for different devices on a multidrop rs485 line using their own read loops and acting somewhat independently from each other.

> remove duplicate requests (issues due to lack of response)

Really I think it's best to just design the app to not make duplicate requests

> Anyhow it is a bigger problem if the APP is not informed about a no-connection state, this goes hand in hand with the APP reconnect_callback at transport layer.

I'm not opposed to adding some notification method, would this just be a client level callback?

> The combination allows the APP to pause and resume requests. 

The app shouldn't normally have to care much beyond using the async API's in a reasonable way, but I could see there being some use cases where a disconnect callback could be handy(such as for triggering some sort of fault notifications/alerts).

> We do not define how the APP works, so we need to protect against patterns we have identified.

Probably a good idea to try and encourage certain design patterns here as relying on a disconnect callback to prevent an unbounded queue seems a little sketchy/unreliable to me, but I could see a disconnect callback or something being a good escape hatch for cases where refactoring may be tricky.

Reply:
> My comments about queuing when not connected stands, the rest of the PR is not blockers but only work to do.

A client should be able to avoid queuing when not connected like this I think:
```python
if client.transport:
    data = await client.read_holding_registers(addr, count, slave=1)
```

Reply:
This is still wrong, the await does not respect the timeout period, so it might wait forever.

Apart from that with a queue it should be buffered.